### ISSUE TITLE
OSCORE fallback use cases

### ISSUE THREAD

**Issue by sbernard31** [eclipse-leshan/leshan#1257]
Using OSCORE, I found a pretty simple use case where the client will not be able to connect to the server anymore.

- I launch a `leshan-server-demo`
- I create oscore config for my device in the UI
- I launch a `leshan-client-demo` using the corresponding OSCORE setting
- simulate a crash (`kill -9 PID` on linux), this way the client doesn't send a deregister.
- then I try to relaunch the client (same OSCORE config)
- and I get a : `Registration failed: UNAUTHORIZED(401) Replay detected.`

@rikard-sics any idea about this ? and how we could handle this ?

**Comment by rikard-sics** [comment]
Spontaneous reaction: This should be solved by making sure the Client uses the Appendix B.2 procedure defined in the OSCORE RFC here:
https://datatracker.ietf.org/doc/html/rfc8613#appendix-B.2
It will re-establish a new OSCORE Security Context, most importantly new Sender/Recipient Keys and a fresh replay window on the server. By doing that the Client restarting from Sender Sequencer Number 0 is not Key and Nonce reuse, and not a replay either. (Note that the Appendix B.2 procedure changes the ID Context but keeps the same Sender/Recipient IDs).

So basically if the client loses mutable state information in the context (like the Sender Sequence Number), it must use this procedure the first time it resumes communication with the server. I believe in the current code the Client will at least use this procedure when it registers to the server.

That being said I need to investigate this case to understand if the Client is trying to use Appendix B.2 in this case. If not, using it should solve this. If the client is trying to use it already, we need to investigate things further.

**Comment by sbernard31** [comment]
> This should be solved by making sure the Client uses the Appendix B.2 procedure defined in the OSCORE RFC here:
https://datatracker.ietf.org/doc/html/rfc8613#appendix-B.2

I don't know what should be done in the code to be sure that appendex B.2 is used ? :thinking: 

> I believe in the current code the Client will at least use this procedure when it registers to the server.

I think it makes sense.
But as you said that _"Note that the Appendix B.2 procedure changes the ID Context "_
What the purpose to set the Context ID in OSCORE object (21) at bootstrap time :thinking:  ?

**Comment by sbernard31** [comment]
@rikard-sics I would really appreciate if we can find a fix (or at least workaround) for this. :pray:

**Comment by sbernard31** [comment]
I try to investigate on this.

> I don't know what should be done in the code to be sure that appendex B.2 is used ? 

Looking at [cf test](https://github.com/eclipse/californium/blob/3.5.0/cf-oscore/src/test/java/org/eclipse/californium/oscore/ContextRederivationTest.java#L144-L147), I saw that maybe the carlifornium API to do that is : 
```java
// Enable context re-derivation functionality (in general)
ctx.setContextRederivationEnabled(true);
// Explicitly initiate the context re-derivation procedure
ctx.setContextRederivationPhase(PHASE.CLIENT_INITIATE);
```

So I tried to add this to [InMemoryOscoreContextDB.deriveContext](https://github.com/eclipse/leshan/blob/278c160c0065a0f80952ad5c7e9fe341cc583c2c/leshan-core-cf/src/main/java/org/eclipse/leshan/core/californium/oscore/cf/InMemoryOscoreContextDB.java#L101-L113) : 
```java
private static OSCoreCtx deriveContext(OscoreParameters oscoreParameters) {
    try {
        OSCoreCtx osCoreCtx = new OSCoreCtx(oscoreParameters.getMasterSecret(), true,
                oscoreParameters.getAeadAlgorithm(), oscoreParameters.getSenderId(),
                oscoreParameters.getRecipientId(), oscoreParameters.getHmacAlgorithm(), 32,
                oscoreParameters.getMasterSalt(), null, 1000);
        osCoreCtx.setContextRederivationEnabled(true);
        // 👇 I try to add this line below 👇
        osCoreCtx.setContextRederivationPhase(PHASE.CLIENT_INITIATE);
        return osCoreCtx;
    } catch (OSException e) {
        LOG.error("Unable to derive context from {}", oscoreParameters, e);
        return null;
    }
}
```

But then I'm not able to connect with OSCORE anymore because of : 
`java.lang.IllegalArgumentException: Internal Leshan operations should always use a null ID Context`.
Raised by [`InMemoryOscoreContextDB.getContext(byte[] rid, byte[] IDContext)`](https://github.com/eclipse/leshan/blob/278c160c0065a0f80952ad5c7e9fe341cc583c2c/leshan-core-cf/src/main/java/org/eclipse/leshan/core/californium/oscore/cf/InMemoryOscoreContextDB.java#L42-L62)

So I tried to remove the check just for testing but this failed with : 
```
org.eclipse.californium.scandium.dtls.cipher.InvalidMacException: MAC validation failed
	at org.eclipse.californium.scandium.dtls.cipher.CCMBlockCipher.decrypt(CCMBlockCipher.java:406)
	at org.eclipse.californium.scandium.dtls.cipher.CCMBlockCipher.decrypt(CCMBlockCipher.java:335)
	at org.eclipse.californium.cose.EncryptCommon.AES_CCM_Decrypt(EncryptCommon.java:134)
	at org.eclipse.californium.cose.EncryptCommon.decryptWithKey(EncryptCommon.java:74)
	at org.eclipse.californium.cose.Encrypt0Message.decrypt(Encrypt0Message.java:138)
	at org.eclipse.californium.oscore.Decryptor.decryptAndDecode(Decryptor.java:140)
	at org.eclipse.californium.oscore.RequestDecryptor.decrypt(RequestDecryptor.java:105)
	at org.eclipse.californium.oscore.ObjectSecurityLayer.prepareReceive(ObjectSecurityLayer.java:104)
	at org.eclipse.californium.oscore.ObjectSecurityLayer.receiveRequest(ObjectSecurityLayer.java:312)
```

Finally, I get to the conclusion that I'm not really able to do that without someone who really well understands OSCORE RFC and OSCORE Californium code, so I give up for now to try to do this alone. 

@rikard-sics, **I'm waiting for your help :pray: !** 

![You're my only hope](https://media.giphy.com/media/l3fZQYk10w0yZ4KfS/giphy.gif)

**Comment by rikard-sics** [comment]
> @rikard-sics, **I'm waiting for your help pray !**

Hmm, it is hard to say what is going wrong from just a look at the output. I will try to take some time to test and see if I can figure out what is happening. Today is a bit tricky but I will try to do it as soon as I can.

**Comment by JaroslawLegierski** [comment]
@rikard-sics 
I reproduced the bug described in this issue and I have first question- the second response from the server is COAP not OSCORE message:
![packets](https://github.com/eclipse-leshan/leshan/assets/13430617/ca0c551d-92f2-4880-be26-344b0d34c115)
However in client log I see the message "Incoming response is NOT OSCORE protected but is expected to be! ":
```
2023-07-28 16:59:15,874 UdpMatcher           [TRACE] received response ACK-4.01   MID=36245, Token=1C0DEB1355706ECA, OptionSet={"Content-Format":"text/plain", "Max-Age":0}, "Replay detected" from UDP(127.0.0.1:5683)  
2023-07-28 16:59:15,876 EndpointContextUtil  [TRACE] udp context receiving, PLAIN: "" == ""  
2023-07-28 16:59:15,876 ObjectSecurityLayer  [INFO] Incoming response is NOT OSCORE protected but is expected to be!  
2023-07-28 16:59:15,876 BlockwiseLayer       [DEBUG] [LWM2M Client-coap://0.0.0.0:0]  received error ACK-4.01   MID=36245, Token=1C0DEB1355706ECA, 
```
The question is : should the 401 response be a plain text COAP or OSCORE protected message ?

**Comment by JaroslawLegierski** [comment]
Based on information from [this](https://github.com/KeithLaiKB/java_test_learn_cf_oscore/blob/16276edeec536645685e63d4cb5d084342fb7f5e/oscore_server_workspace1/com.learn.californium.tryoscore.server/src/main/java/com/learn/californium/tryoscore/server/observe/TestOb2_RederivationEnable.java#L122) and [this ](https://github.com/KeithLaiKB/java_test_learn_cf_oscore/blob/16276edeec536645685e63d4cb5d084342fb7f5e/oscore_client_workspace1/com.learn.californium.tryoscore.client/src/main/java/com/learn/californium/tryoscore/client/observe/TstOb2_Rederivation_CliInit_can2.java#L108) example, I concluded that maybe we need two different implementations of deriveContext method in InMemoryOscoreContextDB class ?:

for server:
```java
    private static OSCoreCtx deriveContext(OscoreParameters oscoreParameters) {
        try {
            OSCoreCtx osCoreCtx = new OSCoreCtx(oscoreParameters.getMasterSecret(), false,
                    oscoreParameters.getAeadAlgorithm(), oscoreParameters.getSenderId(),
                    oscoreParameters.getRecipientId(), oscoreParameters.getHmacAlgorithm(), 32,
                    oscoreParameters.getMasterSalt(), null, 4096);
            osCoreCtx.setContextRederivationEnabled(true);
            return osCoreCtx;
        } catch (OSException e) {
            LOG.error("Unable to derive context from {}", oscoreParameters, e);
            return null;
        }
    }
```	
and for client:
```java
    private static OSCoreCtx deriveContext(OscoreParameters oscoreParameters) {
        try {
            OSCoreCtx osCoreCtx = new OSCoreCtx(oscoreParameters.getMasterSecret(), true,
                    oscoreParameters.getAeadAlgorithm(), oscoreParameters.getSenderId(),
                    oscoreParameters.getRecipientId(), oscoreParameters.getHmacAlgorithm(), 32,
                    oscoreParameters.getMasterSalt(), null, 4096);
            osCoreCtx.setContextRederivationEnabled(true);
            osCoreCtx.setContextRederivationPhase(ContextRederivation.PHASE.CLIENT_INITIATE);
            return osCoreCtx;
        } catch (OSException e) {
            LOG.error("Unable to derive context from {}", oscoreParameters, e);
            return null;
        }
    }
```
Another problem is related to the DEREGISTRATION/REGISTRATION event in EventServlet on server site. As result of context rederivation procedure client is deregistered and registered again and therefore removeContext method in OscoreContextCleaner is cleaning the context. To prevent this, I made the following modification in removeContext method.

```java
    private void removeContext(byte[] rid) {
        OSCoreCtx context = oscoreCtxDB.getContext(rid);
        if (context != null)
            //JL test
            if (!oscoreCtxDB.getContext(rid).getContextRederivationEnabled()) {
                oscoreCtxDB.removeContext(context);
            }
    }
```
I don't know if checking contextRederivationEnabled flag is the best solution - but currently I don't have better idea ...

The next problem concerns the use of IDContext. Leshan operations can use internally  null ID Context also during context rederivation procedure. I set IDContext=null in client and server site and procedure described in OSCORE [appendix-B.2](https://tools.ietf.org/html/rfc8613#appendix-B.2) is [finishing succesfully](https://github.com/eclipse-leshan/leshan/compare/master...JaroslawLegierski:leshan:opl/oscore_tests_1).

This PoC is available [here](https://github.com/JaroslawLegierski/leshan/tree/opl/oscore_tests_1 )

I found in appendix-B.2 that IDContext is used:
- If the client has a unique identifier in some namespace, then that  identifier can be used as ID Context.
- The ID Context may be used to add randomness into new Sender and  Recipient Contexts, see Appendix B.2.
- In the case of group communication [Group-OSCORE], a group  identifier is used as ID Context to enable different security contexts for a server belonging to multiple groups.
 
Does it mean that if we assume in leshan internally IDContext=null can we lose some OSCORE functionality ? :worried:

**Comment by JaroslawLegierski** [comment]
@sbernard31  Do you have any LwM2M device with OSCORE that you could recommend for testing ? I'm looking for device with properly/fully implemented OSCORE that I could to use for testing my changes in Leshan code.

**Comment by sbernard31** [comment]
No idea, I didn't play so much with OSCORE.
At first sight : 
 - [libcoap](https://github.com/obgm/libcoap#package-contents) seems to support it but this is **CoAP only**.
 - [anjay](https://avsystem.github.io/Anjay-doc/CommercialFeatures.html) also seems to support it but in the **commercial version only**.

**Comment by JaroslawLegierski** [comment]
I prepared next PoC oriented on Appendix.B.2 implementation. In deriveContext method in InMemoryOscoreContextDB class I'm using IDContext!=null.

Please find here full implementation [here](https://github.com/eclipse-leshan/leshan/compare/master...JaroslawLegierski:leshan:opl/oscore_tests_2). 

Unfortunately to working this implementation properly following modification is mandatory on Califorium site [(class ContextRederivation line 340)](https://github.com/eclipse-californium/californium/compare/main...JaroslawLegierski:californium-test:opl/oscore_appx_b2):

``` java
			if (contextID == null|| (Arrays.equals(contextID, ctx.getIdContext()))
                                  &&!ctx.getContextRederivationEnabled())  //<- my modification
                         {
				return ctx;
			}
```
@rikard-sics  How do you see the possibility adding such modification to OSCORE library in Cf ?

**Comment by rikard-sics** [comment]
> @rikard-sics What is Your opinion about such procedure of context derivation initialization ?

Let me take a look during the beginning of next week, and I can follow up in this thread.

**Comment by JaroslawLegierski** [comment]
Reading  carefully [OSCORE Appendix B.2](https://www.rfc-editor.org/rfc/rfc8613.html#page-71) I found that:
_The server sends a 4.01 (Unauthorized) response protected with the second security context, containing R2 wrapped in a CBOR bstr as  'kid context', and caches R2._


> In this PoC leshan-client-demo is starting [OSCORE Appendix B.2 procedure](https://www.rfc-editor.org/rfc/rfc8613.html#page-71) based on 4.01 (Unauthorized) response from the server.
> 
> ![b2_with_401](https://user-images.githubusercontent.com/13430617/270658872-04557e91-7a99-43e1-91a9-5e407559b355.png)
> 
> @rikard-sics What is Your opinion about such procedure of context derivation initialization ?

@rikard-sics Does this mean that mentioned above the 401 Unauthorized message should be also OSCORE encoded?

**Comment by sbernard31** [comment]
@rikard-sics do you still plan to work on OSCORE support in Leshan ? We didn't get any news from you since a very long time.

**Comment by rikard-sics** [comment]
> [@rikard-sics](https://github.com/rikard-sics) do you still plan to work on OSCORE support in Leshan ? We didn't get any news from you since a very long time.

Yes, I do still plan to work on OSCORE in Leshan. It would be great if we can get to a situation where OSCORE support is no longer experimental (if feasible). We could discuss further regarding what to do in order to accomplish that. But I suppose working on this issue is a good place to start?

In the OSCORE code in Californium I have JUnit tests for Appendix B.2 and there things work well. So the plan could be to understand why things are going wrong when running OSCORE Appendix B.2 integrated with the Leshan code.

**Comment by sbernard31** [comment]
> I do still plan to work on OSCORE in Leshan

Great to hear.

>  It would be great if we can get to a situation where OSCORE support is no longer experimental (if feasible).

👍 

>  But I suppose working on this issue is a good place to start?

I totally agree.

> In the OSCORE code in Californium I have JUnit tests for Appendix B.2 and there things work well. So the plan could be to understand why things are going wrong when running OSCORE Appendix B.2 integrated with the Leshan code.

👍

**Comment by rikard-sics** [comment]
Alright, taking more time to look into this, this is what I concluded.

What is reported in the first message in this thread is completely expected. That is, if the client restarts and re-uses the same OSCORE context then the server will detect incoming requests as replays. This is because the client starts its sender sequence number over from 0, with the exact same context and its parameters in terms of Sender/Recipient keys, Sender ID and ID Context. The replay window on the server associated with that context has already marked previously sent requests from the client as received. (So, a new context needs to be derived using Appendix B.2.)

Due to this the LWM2M 1.1 Transport Bindings document defines the use of Appendix B.2 for OSCORE. Another reason is that the client will avoid performing key and nonce re-use, which would be unsafe.

**Previous pull request to add use of Appendix B.2 to Leshan**

In 2021 I created a PR to add the use of OSCORE Appendix B.2 in Leshan, which was merged. It also explains the problem I summarized above in greater detail.
See that PR here: https://github.com/eclipse-leshan/leshan/pull/1004

However, in the latest code of Leshan, no code related to Appendix B.2 is present. At the time of the PR only a few lines were necessary to add support for Appendix B.2. But obviously the Leshan code has changed since then and it seems we need more extensive additions to add use of Appendix B.2 to the current codebase.


**Steps forward**

In my mind this is what should happen: Every time the client creates an OSCORE context (since it just started, since it just bootstrapped, or since it restarted) it should mark that context with:
`ctx.setContextRederivationPhase(PHASE.CLIENT_INITIATE);`

Also any OSCORE context on the client, server or bootstrap server must have Appendix B.2 enabled using:
`ctx.setContextRederivationEnabled(true);`

What we want to achieve is that the first time the client sends a request to the server or bootstrap server, that request will initiate Appendix B.2. If the client restarts, then that must happen again for the first request sent from the client to the server or bootstrap server.

Now I understand there are some complications with this, but I need to experiment with the code to understand exactly what those are.


**Regarding unprotected 4.01 responses**

> However in client log I see the message "Incoming response is NOT OSCORE protected but is expected to be! ":

That response which contains "Replay detected" should never happen since Appendix B.2 should be ran before that.

**Comment by sbernard31** [comment]
Thx @rikard-sics for analysis.

I see some  ` ctx.setContextRederivationEnabled(true);` in the code.
But I didn't see any `ctx.setContextRederivationPhase(PHASE.CLIENT_INITIATE);` (maybe I missed something)

I will try to investigate on that point.

**Comment by rikard-sics** [comment]
> Thx [@rikard-sics](https://github.com/rikard-sics) for analysis.
> 
> I see some ` ctx.setContextRederivationEnabled(true);` in the code. But I didn't see any `ctx.setContextRederivationPhase(PHASE.CLIENT_INITIATE);` (maybe I missed something)
> 
> I will try to investigate on that point.

Sure! I looked through the latest master and couldn't find any mention of ` ctx.setContextRederivationEnabled(true);` or `ctx.setContextRederivationPhase(PHASE.CLIENT_INITIATE);`  but perhaps I missed some instance.

I will have a test with the code also, setting things up to use Appendix B.2 and seeing what happens (and what errors may occur).

**Comment by rikard-sics** [comment]
Alright it will take me some time to get things set up to develop. I have Java 17, and a compatible version of Maven and Eclipse, so I can import and run the LeshanDemoServer. But I'm running into some issues with node and npm so I can't actually use the webserver.

**Adding context by URI**

Regarding this part:
```
// TODO OSCORE don't know if I should add by uri here ?
// see : https://github.com/eclipse/leshan/pull/1212#discussion_r830937966
// For now we don't add it because of :
// https://github.com/eclipse/leshan/pull/1232#discussion_r841851488
// super.addContext(uri, osCoreCtx);
```

Are contexts ever added by URI? For outgoing requests the context is retrieved based on the URI, so if contexts are not associated with URIs, the server will not find a context when sending requests to the client. I recall adding code in the past to update the context database upon reception of the first request from the client (when the server learns the hostname & port of the client).

But I see you have discussions in other issues about problems with associating contexts with URIs, I'll check out those discussions to refresh my memory.

**Regarding OSCORE context retrieval**

Just to mention that when retrieving OSCORE contexts from the OSCORE context database (_HashMapCtxDB_), the ID Context is not needed if there is **only a single OSCORE context found for the specified RID**. So if Leshan restricts itself to never allowing multiple contexts with the same RID, then contexts can be retrieved using only `super.getContext(rid)`.

So contexts would get an ID context after running Appendix B.2, but it would be irrelevant and not used for context retrieval. Of course the problem is that enforcing that all contexts must have unique RIDs may be infeasible. Just a note if that's useful.

**Promising approach**

Otherwise from what I see in the code, it seems most relevant things are concentrated in _InMemoryOscoreContextDB_. I think the direction @JaroslawLegierski was working on [here](https://github.com/eclipse-leshan/leshan/issues/1257#issuecomment-1693536195) looked promising. Having a separate method for deriving contexts for clients makes sense. Because then that method could include:
`osCoreCtx.setContextRederivationPhase(PHASE.CLIENT_INITIATE);`

I wanted to comment on this part:

> In this PoC leshan-client-demo is starting OSCORE Appendix B.2 procedure based on 4.01 (Unauthorized) response from the server.

The client should initiate Appendix B.2 immediately when it sends its first request to the server. (That first request should be Request 1 of the Appendix B.2 procedure). The intent of setting `PHASE.CLIENT_INITIATE` is that the next following request the client sends will trigger initiation of Appendix B.2 automatically.

So I'm not sure about why the design choice of waiting for a "Replay detected" error message before initiating Appendix B.2. Excluding that, I believe that approach should work and I would be curious to see what happens if it's tried. Basically if equivalent code as in PR #1004 were applied to the current codebase (making sure to only set the phase on the client), what would happen?

Anyways, I hope to be able to start experimenting a bit myself this week if I can finish getting the dev environment set up.

**Comment by sbernard31** [comment]
> But I'm running into some issues with node and npm so I can't actually use the webserver.

Do not hesitate to open a dedicated issue if needed.

> Are contexts ever added by URI?

Currently, I think it is never added. 

> For outgoing requests the context is retrieved based on the URI, so if contexts are not associated with URIs, the server will not find a context when sending requests to the client.

The store allow to strongly associate URI <=> RID, so when we search by URI, we find corresponding RID then we search by RID under the hood.

> But I see you have discussions in other issues about problems with associating contexts with URIs, I'll check out those discussions to refresh my memory.

Yep, hard to get all those in mind again :)

> So if Leshan restricts itself to never allowing multiple contexts with the same RID, then contexts can be retrieved using only super.getContext(rid).

Maybe it's ok to restrict that in a first time. 

-----------------------------

I tried to play a bit with current code (especially trying to add `osCoreCtx.setContextRederivationPhase(PHASE.CLIENT_INITIATE);` and I didn't succeed to make it work.

I didn't understand OSCORE so well, I know californium code a bit more and I know well Leshan code and I didn't succeed to use OSCORE CF API correctly.

Maybe it's me but also maybe we have a problem with current API or at least we need a more user-friendly API on top of current one ? 

This makes me think that maybe the priority is more to define that API (so more a specification work) than trying to fix this issue ? 
I will maybe create a dedicated issue about it 🤔

**Comment by rikard-sics** [comment]
> I tried to play a bit with current code (especially trying to add `osCoreCtx.setContextRederivationPhase(PHASE.CLIENT_INITIATE);` and I didn't succeed to make it work.

I managed to get the development environment set up and was working on this yesterday and made progress. I will send an update later today about that and other points

**Comment by rikard-sics** [comment]
> Do not hesitate to open a dedicated issue if needed.

I worked it out. It was mainly a problem of versioning.


> The store allow to strongly associate URI <=> RID, so when we search by URI, we find corresponding RID then we search by RID under the hood.

I see! Yeah makes sense


> Maybe it's ok to restrict that in a first time.

I would say so. At least for now having a restriction of not allowing multiple OSCORE contexts with the same RID can be a good solution. In most situations ensuring that through configuration should be feasible.


> I didn't understand OSCORE so well, I know californium code a bit more and I know well Leshan code and I didn't succeed to use OSCORE CF API correctly.
> 
> Maybe it's me but also maybe we have a problem with current API or at least we need a more user-friendly / user-oriented API on top of current one ?
> 
> This makes me think that maybe the priority is more to define that API (so more a specification work) than trying to fix this issue ? I will maybe create a dedicated issue about that 🤔

I didn't get much feedback on the API from a developer or user point of view, and it kind of grew organically. However, I would say that if you look at the OSCORE [client](https://github.com/eclipse-californium/californium/blob/main/cf-oscore/src/test/java/org/eclipse/californium/oscore/HelloWorldClient.java) and [server](https://github.com/eclipse-californium/californium/blob/main/cf-oscore/src/test/java/org/eclipse/californium/oscore/HelloWorldServer.java) examples, not so much code is needed to set these up.

As for Appendix B.2, the intent is to have a simple interface and require only the two lines `ctx.setContextRederivationEnabled(true);` & `ctx.setContextRederivationPhase(PHASE.CLIENT_INITIATE);` on the client and one line on the server. In that case I'd say the problem is more that it's not working as expected in the context of Leshan rather than the API.

However, I'm sure there's room for improvement. Perhaps something like an `OscoreCtxBuilder` could be useful. Feel free to create an issue if you have some ideas.

**Comment by sbernard31** [comment]
About API, I hope you didn't take that the wrong way 🙇‍♀️. I want to insist on my _"maybe"_, this was just an assumption 🤷.

> it kind of grew organically

Yep I understand. 

> As for Appendix B.2, the intent is to have a simple interface and require only the two lines ctx.setContextRederivationEnabled(true); & ctx.setContextRederivationPhase(PHASE.CLIENT_INITIATE); on the client and one line on the server

Yep maybe I'm wrong but I see `HashMapCtxDB` and `OSCoreCtx` as a pretty low level API.
And my idea is that maybe we need a more high level API oriented to user needs and focus on experimented users and most common use-cases.  (but we probably already begin to talk a bit about that with #920, #1231, #1230)


>  Feel free to create an issue if you have some ideas.

Yep, I will do that to try clarify that for me and have a place to discuss about that.

**Comment by rikard-sics** [comment]
> About API, I hope you didn't take that the wrong way 🙇‍♀️. I want to insist on my _"maybe"_, this was just an assumption 🤷.

No not at all, maybe it didn't come through in the text but I'm happy to take any feedback! :thumbsup:

I just wanted to highlight that doing "basic" things with OSCORE should be fairly simple. However, of course I risk becoming somewhat blind to my code, and it's good to hear an outsider's perspective.

> > As for Appendix B.2, the intent is to have a simple interface and require only the two lines ctx.setContextRederivationEnabled(true); & ctx.setContextRederivationPhase(PHASE.CLIENT_INITIATE); on the client and one line on the server
> 
> Yep maybe I'm wrong but I see `HashMapCtxDB` and `OSCoreCtx` as a pretty low level API.

Yeah for sure those can be considered low level APIs. As I mentioned I would be interested in having something like an 
OscoreContexBuilder so you can do:
```
OscoreContextBuilder oscBuilder = new OscoreContextBuilder();
oscBuild.setMasterSecret("0102030405060708090a0b0c0d0e0f10");
oscBuild.setRecipientId("00");
oscBuild.setSenderId("01");
OSCoreCtx ctx = oscBuilder.build();
== OR ==
private final static HashMapCtxDB db = new HashMapCtxDB();
db.addContext(uri, "0102030405060708090a0b0c0d0e0f10", "00", "01");
```
(Maybe not using strings directly but StringUtil.hex2ByteArray)

In that way the developer/user doesn't have to use the complicated constructor of OSCoreCtx.

For context rederivation with Appendix B.2, I'm sure it can be improved too but I don't have any concrete ideas.

> (I read again https://github.com/eclipse-leshan/leshan/issues/920, https://github.com/eclipse-leshan/leshan/issues/1231, https://github.com/eclipse-leshan/leshan/issues/1230 and most High level API concerns are already there. Sorry for the noise)

No worries, feel free to create a new issue if you prefer. Regardless I will also check those three issues to refresh my memory.

**Comment by rikard-sics** [comment]
So coming back to the problem with Appendix B.2. I've been working on this and this is my current status.

Initially I noticed this:

![Image](https://github.com/user-attachments/assets/5ab8d657-7515-4586-94c2-168945c13a82)

In the fourth message, which is Response 2 of the Appendix B.2 procedure, the ID Context value in the OSCORE option of the message should be empty. However, in the capture it has a value set. This, and looking at the textual output with TRACE debugging, made me notice that the peers are not going through the phases of the context rederivation as I would expect. In fact it seemed that one of the peers ended up using the wrong OSCORE context during the execution of the procedure.

My understanding is that this problem was caused by an interaction of two pieces of code here:
```
 OSCoreCtx osCoreCtx = super.getContext(rid, IDContext);

        if (osCoreCtx == null && IDContext != null) {
            throw new IllegalArgumentException("Internal Leshan operations should always use a null ID Context");
        }

        // if nothing found
        if (osCoreCtx == null) {
            // try to derive new context from OSCORE parameter in OSCORE Store
            OscoreParameters params = store.getOscoreParameters(rid);
            if (params != null) {
                osCoreCtx = deriveContext(params);
                // add new new context in local DB
                super.addContext(osCoreCtx);
            }
        }
```

The context is retrieved using the ID Context which becomes problematic when Appendix B.2 is used and the ID Context changes. This in turn would have been fine, if it was not for the fact that a new context is derived automatically if no context is found. Inside the ContextRederivation class for Appendix B.2 there are places where the context is expected to be null, that is no context was found for this RID and ID Context, in which case the Appendix B.2 logic will instead retrieve the context only by RID. It seems the automatic creation of new contexts disrupted this functionality.

**Updates to fix the functionality**

I found a way to update the current code so that the Appendix B.2 functionality works. This relies on having a separate context creation method for clients and not using the ID Context for context retrieval. (This means adding the restriction we discussed of not having multiple OSCORE contexts with the same RID).

You can see the diff of my updates here:
https://github.com/eclipse-leshan/leshan/compare/master...rikard-sics:leshan:oscore_appb2?expand=1#diff-35ec42d9c87b035f92d28fa23f731a45df7c0a2839a8f50d11dca5a23b1cb4a3

Let's discuss more, but I can continue working on this and make it into a PR.

**Comment by sbernard31** [comment]
@rikard-sics I just think that I should be able to create the PR using your code and adding my test.

If it can help you and enable you to move forward on other points, I'll be happy to do so. 🙂

**Comment by rikard-sics** [comment]
> [@rikard-sics](https://github.com/rikard-sics) I just think that I should be able to create the PR using your code and adding my test.
> 
> If it can help you and enable you to move forward on other points, I'll be happy to do so. 🙂

I'm working to clean up the code and finalize the PR, so there shouldn't be an issue. I can create the PR during the afternoon.

**Comment by sbernard31** [comment]
#1694 is now integrated in `master`

@rikard-sics thx a lot for that. 🙏 

I think we can close this issue right ?

I don't known what should be the next topic : 
 - https://github.com/eclipse-leshan/leshan/issues/1230
 - https://github.com/eclipse-leshan/leshan/issues/1231

**Comment by rikard-sics** [comment]
> [#1694](https://github.com/eclipse-leshan/leshan/pull/1694) is now integrated in `master`
> [@rikard-sics](https://github.com/rikard-sics) thx a lot for that. 🙏

That's great!

> I think we can close this issue right ?

Yep, sounds good to me.

> I don't known what should be the next topic :

I'd propose going for #1230 as it's related to PR #1694, and also smaller in scope than #1231.

I'll follow up with a post in #1230 before the end of the day.
